hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/upgrade/ReconLayoutVersionManager.java

```diff
@@ -0,0 +1,142 @@
+   * @return The current Metadata Layout Version (MLV).
+   */
+  private int determineMLV() {
+    return schemaVersionTableManager.getCurrentSchemaVersion();
+  }
+
+  /**
+   * Determines the Software Layout Version (SLV) based on the latest feature version.
+   * @return The Software Layout Version (SLV).
+   */
+  private int determineSLV() {
+    return Arrays.stream(ReconLayoutFeature.values())
+        .mapToInt(ReconLayoutFeature::getVersion)
+        .max()
+        .orElse(0); // Default to 0 if no features are defined
+  }
+
+  /**
+   * Finalizes the layout features that need to be upgraded, by executing the upgrade action for each
+   * feature that is registered for finalization.
+   */
+  public void finalizeLayoutFeatures() {
+    // Get features that need finalization, sorted by version
+    List<ReconLayoutFeature> featuresToFinalize = getRegisteredFeatures();
+
+    for (ReconLayoutFeature feature : featuresToFinalize) {
+      try {
+        // If the feature is INITIAL_VERSION, skip executing any action and just update the schema version
+        if (feature == ReconLayoutFeature.INITIAL_VERSION) {
+          updateSchemaVersion(0);
+          LOG.info("INITIAL_VERSION feature processed by setting schema version to 0.");
+          continue;
+        }
+
+        // Fetch only the FINALIZE action for the feature
+        Optional<ReconUpgradeAction> action = feature.getAction(ReconUpgradeAction.UpgradeActionType.FINALIZE);
+        if (action.isPresent()) {
+          // Execute the upgrade action & update the schema version in the DB
+          action.get().execute();
+          updateSchemaVersion(feature.getVersion());
+          LOG.info("Feature versioned {} finalized successfully.", feature.getVersion());
+        }
+      } catch (Exception e) {
+        LOG.error("Failed to finalize feature {}: {}", feature.getVersion(), e.getMessage());
+        break;
+      }
+    }
+  }
+
+  /**
+   * Returns a list of ReconLayoutFeature objects that are registered for finalization.
+   */
+  protected List<ReconLayoutFeature> getRegisteredFeatures() {
+    List<ReconLayoutFeature> allFeatures =
+        Arrays.asList(ReconLayoutFeature.values());
+
+    LOG.info("Current MLV: {}. SLV: {}. Checking features for registration...", currentMLV, currentSLV);
+
+    List<ReconLayoutFeature> registeredFeatures = allFeatures.stream()
+        .filter(feature -> feature.getVersion() > currentMLV)
+        .sorted((a, b) -> Integer.compare(a.getVersion(), b.getVersion())) // Sort by version in ascending order
+        .collect(Collectors.toList());
+
+    return registeredFeatures;
+  }
+
+  /**
+   * Updates the Software Layout Version (SLV) in the database after finalizing a feature.
+   * @param newVersion The new Software Layout Version (SLV) to set.
+   */
+  private void updateSchemaVersion(int newVersion) {
+    // Logic to update the MLV in the database
+    this.currentMLV = newVersion;
```

Comment:
setting currentMLV can be after db update. if DB have some exception, this flag is not required to be updated.